src/libcollections/string.rs

```diff
@@ -823,6 +832,18 @@ impl<'a> Extend<&'a str> for String {
     }
 }
 
+#[stable(feature = "extend_string", since = "1.4.0")]
+impl Extend<String> for String {
+    fn extend<I: IntoIterator<Item=String>>(&mut self, iterable: I) {
+        let iterator = iterable.into_iter();
+        let (lower_bound, _) = iterator.size_hint();
+        self.reserve(lower_bound);
```

Comment:
Does the observation from #27986 apply here, too? (i.e. the strings from the iterator might be empty, hence it is better to keep the code simpler and trust the reallocation strategy)